### PR TITLE
Fixes #3796: Deploy command continues despite composer failure

### LINKED ISSUE #3796: DT-669: Deploy command continues despite composer failure
**Describe the bug**
The `blt artifact:deploy` command will deploy artifacts even if a step in the build process fails. The specific example used is a patch that fails to apply to a contrib module, while fails the composer step, which fails to generate the `vendor/autoload.php`, which causes the deployed artifact to crash the environment with an error similar to this:

```
Fatal error: Uncaught Exception: Could not locate autoload.php. cwd is /mnt/www/html/myacsf.01dev/docroot; __DIR__ is /mnt/www/html/myacsf01dev/vendor/drush/drush in /mnt/www/html/myacsf01dev/vendor/drush/drush/drush.php:54
```

**To Reproduce**

Steps to reproduce the behavior, ideally starting from a fresh install of BLT:
1. Run `composer create-project acquia/blt-project --no-interaction`
2. Add a contrib module, here we'll use this module as an example: `composer require drupal/svg_image_field`
3. Add a patch for that file that we know doesn't apply in the composer.json:
    ```
    "patches": {
        "drupal/svg_image_field": {
          "This doesn't apply": "https://www.drupal.org/files/issues/entity-browser-view-context-2865928-14.patch"
        },
    ```
4. Commit the changes
5. Run `blt artifact:deploy`
6. Notice that it succeeds in pushing commits to Acquia. Specifics to observe in the output are:
    a. Error that the patch does not apply:
        ```
        [Exception]
        Cannot apply patch This doesn't apply (https://www.drupal.org/files/issues/entity-browser-view-context-2865928-14.patch)!
        ```
    b. The ExecStack attempts to exit the script:
        ```
        [ExecStack]  Exit code 1  Time 16.434s
        ```
    c. The script continues:
        ```
        Sanitizing artifact...
        ```

**Expected behavior**
At step 6 in the steps above the expected behavior is that if the patch does not apply the entire `artifact:deploy` script fails and exits. It should NOT push commits to Acquia.

**Detailed error output**
```
Commit message is set to Automated commit by Travis CI for Build .
Branch is set to develop-build.
Preparing artifact directory...
(lost of output)
  - Applying patches for drupal/svg_image_field
    https://www.drupal.org/files/issues/entity-browser-view-context-2865928-14.patch (This doesn't apply)
   Could not apply patch! Skipping. The error was: Cannot apply patch https://www.drupal.org/files/issues/entity-browser-view-context-2865928-14.patch


  [Exception]
  Cannot apply patch This doesn't apply (https://www.drupal.org/files/issues/entity-browser-view-context-2865928-14.patch)!


install [--prefer-source] [--prefer-dist] [--dry-run] [--dev] [--no-dev] [--no-custom-installers] [--no-autoloader] [--no-scripts] [--no-progress] [--no-suggest] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-            authoritative] [--apcu-autoloader] [--ignore-platform-reqs] [--] [<packages>]...

[ExecStack]  Exit code 1  Time 16.434s
Sanitizing artifact...
[info] Find Drupal core text files...
(more BLT output)
[ExecStack] Done in 5.343s
Pushing artifact to git.remotes...
[ExecStack] git push 470bf68fce4b28e6f615e854987d6dc3 develop-build
[ExecStack] Running git push 470bf68fce4b28e6f615e854987d6dc3 develop-build in /tmp/deploy
Counting objects: 117, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (104/104), done.
Writing objects: 100% (117/117), 56.54 KiB | 0 bytes/s, done.
Total 117 (delta 81), reused 0 (delta 0)
To myacsf@svn-2398.enterprise-g1.hosting.acquia.com:myacsf.git
   11d586f..f8f9bc2  develop-build -> develop-build
[ExecStack] Done in 2.455s
01:35 total time elapsed.

set +v
```

**BLT doctor output**
```vagrant@myacsf:/var/www/myacsf$ blt doctor
+---------------------------+--------------------------------------------------------------+
| Property                  | Value                                                        |
+---------------------------+--------------------------------------------------------------+
| %paths.%root              | /var/www/myacsf/docroot                                      |
| %paths.%site              | sites/default                                                |
| %paths.%modules           | sites/all/modules                                            |
| %paths.%themes            | sites/all/themes                                             |
| %paths.%config-sync       | ../docroot/profiles/custom/sfprofile/config/sync             |
| %paths.%config-vcs        | ../docroot/profiles/custom/sfprofile/config/sync             |
| %paths.%files             | sites/g/files/kebobs123/files                                |
| %paths.%temp              | /tmp                                                         |
| %paths.%private           | /var/www/myacsf/files-private/kebobs123                      |
| admin-theme               | seven                                                        |
| alias-searchpaths.0       | /var/www/myacsf/drush/sites                                  |
| blt-version               | 9.2.8                                                        |
| bootstrap                 | Successful                                                   |
| composer-version          | Composer version 1.8.5 2019-04-09 17:46:47                   |
| config-sync               | ../docroot/profiles/custom/sfprofile/config/sync             |
| db-driver                 | mysql                                                        |
| db-hostname               | 127.0.0.1                                                    |
| db-name                   | myacsf                                                       |
| db-password               | myacsf                                                       |
| db-port                   | 3306                                                         |
| db-status                 | Connected                                                    |
| db-username               | myacsf                                                       |
| drupal-settings-file      | sites/default/settings.php                                   |
| drupal-version            | 8.7.2                                                        |
| drush-alias-files.0       | /var/www/myacsf/drush/sites/mysite.site.yml                  |
|                           | yml                                                          |
| drush-cache-directory     | /home/vagrant/.drush/cache                                   |
| drush-conf.0              | /var/www/myacsf/vendor/drush/drush/drush.yml                 |
| drush-conf.1              | /var/www/myacsf/drush/drush.yml                              |
| drush-conf.2              | /var/www/myacsf/docroot/sites/default/local.drush.yml        |
| drush-script              | /var/www/myacsf/vendor/bin/drush                             |
| drush-temp                | /tmp                                                         |
| drush-version             | 9.5.2                                                        |
| files                     | sites/g/files/kebobs123/files                                |
| install-profile           | sfprofile                                                    |
| modules                   | sites/all/modules                                            |
| php-bin                   | /usr/bin/php7.1                                              |
| php-conf.0                | /etc/php/7.1/cli/php.ini                                     |
| php-os                    | Linux                                                        |
| private                   | /var/www/myacsf/files-private/kebobs123                      |
| root                      | /var/www/myacsf/docroot                                      |
| site                      | sites/default                                                |
| stacks.drupal-vm.inited   | true                                                         |
| stacks.dev-desktop.inited | false                                                        |
| temp                      | /tmp                                                         |
| theme                     | bartik                                                       |
| themes                    | sites/all/themes                                             |
| uri                       | http://myacsf.local                                          |
+---------------------------+--------------------------------------------------------------+
+--------------------------------------+-----------------------------------------------------------+
| Check                                | Problem                                                   |
+--------------------------------------+-----------------------------------------------------------+
| NodeCheck:checkNodeVersionFileExists | Neither .nvmrc nor .node-version file found in repo root. |
+--------------------------------------+-----------------------------------------------------------+
```

**System information**
* Operating system type: Debian
* Operating system version: Ubuntu 16.04.5 LTS
* BLT version: 9.2.8

**Additional context**
In practice, this appears to happen composer requests a patch from drupal.org and gets a non-200 or corrupted response which results in composer not being able to apply the patch. It seems like it would be rare, but it has happened twice in the last week:

* Fri Aug 2 14:00:45 2019 +0000
* Sat Aug 3 01:29:46 2019 +0000

### PR DESCRIPTION
**Pull request by nhmbounteous**
Fixes #3796 
--------

Changes proposed
---------
Added logic check to see if the `composer install` command was successful, and if not interrupt the deployment to prevent corrupt environment code

Steps to replicate the issue
----------
1. Create a semi-corrupt compsoer.json
2. Example, apply incorrect patch to module
```
"patches": {
    "drupal/svg_image_field": {
      "This doesn't apply": "https://www.drupal.org/files/issues/entity-browser-view-context-2865928-14.patch"
    },
```
3. Run `blt artifact:deploy`
4. The deployment will display the error `[Exception] Cannot apply patch This doesn't apply (https://www.drupal.org/files/issues/entity-browser-view-context-2865928-14.patch)!`
5. The error will be ignored and cause a broken artifact to be pushed to the environment. 

Previous (bad) behavior, before applying PR
----------
Corrupt environment code on deployment and ignored the composer error

Expected behavior, after applying PR and re-running test steps
-----------
An exception is now thrown and the process is stopped, preventing a corrupted artifact deployment

### CODE REVIEW

File: src/Robo/Commands/Deploy/DeployCommand.php [hunk 1]
```diff
@@ -488,10 +488,13 @@ protected function composerInstall() {
     if ($this->ignorePlatformReqs) {
       $command .= ' --ignore-platform-reqs';
     }
-    $this->taskExecStack()->exec($command)
+    $execution_result = $this->taskExecStack()->exec($command)
       ->stopOnFail()
       ->dir($this->deployDir)
       ->run();
+    if(!$execution_result->wasSuccessful()) {
+      throw new BltException("Composer Install failed, please review and check your composer.json file.");
```
Comment (josephdpurcell) [inline src/Robo/Commands/Deploy/DeployCommand.php]: Is there a way to pass the $exception_result message as the exception?

Comment (nhmbounteous) [reply src/Robo/Commands/Deploy/DeployCommand.php]: I'm not sure if that would look the best, due to the error being displayed already above the exception output. Also, it does appear that this is the suggested format and there might be a re-implementation on how these exceptions are going to be captured per danepowell last comment.

Comment (josephdpurcell) [reply src/Robo/Commands/Deploy/DeployCommand.php]: Ok, let's adjust messaging to account for the scenario where the composer install failed due to network issue.

What about:

```
Composer install failed, please check the output for details.
```

**Comment by danepowell** [comment]
@alexxed we are checking on that with the Robo maintainers here: https://github.com/consolidation/Robo/issues/891

But for now, it looks like no, Robo does not throw exceptions, it just provides an exit code.

**Pull request merged**